FIrst draft hydra head (#1469)
input-output-hk/hydra: docs/docs/tutorial/index.md

> # Open a head on testnet

Reviewer: ```suggestion
# Opening a head on testnet
```

This tutorial demonstrates how to use `hydra-node` on the `preprod` Cardano network to open a layer-two state channel between two participants using the Hydra Head protocol. We'll also employ [Mithril](https://mithril.network) for rapid node setup.

> This setup follows the [Basic Hydra Head](/topologies/basic) topology, establishing the "green" Hydra Head between `X` and `Y` as shown below:

Reviewer: ```suggestion
This tutorial demonstrates how to use `hydra-node` on Cardano's `pre-production` testing environment to open a layer 2 state channel between two participants using the Hydra Head protocol. We'll also use [Mithril](https://mithril.network) for fast node bootstrapping.

This setup follows the [basic Hydra Head](/topologies/basic) topology, establishing the 'green' Hydra Head between `X` and `Y` as shown below:
```

![](../../topologies/basic/basic-hydra-head.jpg)

## Prerequisites

- [ ] Terminal access to an internet-connected machine.
- [ ] Either:
  - [ ] Another participant following this tutorial (recommended), or
  - [ ] Access to two such machines.

> - [ ] 100 tADA per participant in a wallet on the `preprod` network.

Reviewer: ```suggestion
- [ ] Terminal access to an internet-connected machine
- [ ] Either:
  - [ ] Another participant following this tutorial (recommended), or
  - [ ] Access to two such machines
- [ ] 100 test ada per participant in a wallet on the `pre-production` network.
```

> ## Step 0: Installation

Reviewer: ```suggestion
## Step 0. Installation
```

> Required tools; This tutorial assumes the following tools are available on your system:

Reviewer: ```suggestion
**Required tools** 

This tutorial assumes the following tools are available on your system:
```

- [ ] `curl`
- [ ] `tar`
- [ ] `unzip`
- [ ] [`websocat`](https://github.com/vi/websocat)

> - [ ] [`jq`](https://jqlang.github.io/jq/)

Reviewer: ```suggestion
- [ ] [`jq`](https://jqlang.github.io/jq/).
```

<details>
<summary>Other installation options</summary>

Explore alternative methods for acquiring and running the Cardano, Mithril, and Hydra nodes, depending on your environment:

- Docker containers are published regularly.
- Some projects provide system-level packages for installation and/or pre-built binaries for various platforms.

> - Building from source is always an option.

Reviewer: ```suggestion
- Docker containers are published regularly
- Some projects provide system-level packages for installation and/or pre-built binaries for various platforms
- Building from source is always an option.
```

Please check out each project's GitHub pages for more options.

</details>

> ## Step 1: Connect to Cardano

Reviewer: ```suggestion
## Step 1. Connect to Cardano
```

> The Hydra Head protocol requires a connection to the Cardano layer-one network to verify and post protocol transactions in a trustless way. Start by setting up a `cardano-node` on the public testnet. With Mithril, you can skip synchronizing the whole history and get started quickly.

Reviewer: ```suggestion
The Hydra Head protocol requires a connection to the Cardano layer 1 network to verify and post protocol transactions in a trustless way. Start by setting up a `cardano-node` on the public testnet. With Mithril, you can skip synchronizing the whole history and get started quickly.
```

> Download the latest blockchain snapshot using `mithril-client` configured for the `preprod` network:

Reviewer: ```suggestion
Download the latest blockchain snapshot using `mithril-client` configured for the `pre-production` network:
```

<details>
<summary>Bash auto-completion</summary>

Enhance your `bash` experience with auto-completion for `cardano-cli`:

```shell
source <(cardano-cli --bash-completion-script cardano-cli)
```

</details>

For more detailed instructions on setting up a `cardano-node` using Mithril, refer to [this guide](https://mithril.network/doc/manual/getting-started/bootstrap-cardano-node).

> ## Step 2: Prepare keys and funding

Reviewer: ```suggestion
## Step 2. Prepare keys and funding
```

This tutorial assumes a minimal setup with two participants named `alice` and `bob`. Depending on your setup, choose an appropriate role or perform actions for both roles if working alone.

> First, generate Cardano key pairs and addresses for both participants to identify the `hydra-node` and manage funds on layer one:

Reviewer: ```suggestion
First, generate Cardano key pairs and addresses for both participants to identify the `hydra-node` and manage funds on layer 1:
```

</TabItem>
</Tabs>

> Next, fund these addresses. If you have tADA on `preprod`, distribute it to the newly created addresses:

Reviewer: ```suggestion
Next, fund these addresses. If you have test ada on the `pre-production` network, distribute it to the newly created addresses:
```

:::info Where to get funds

> In case you have no tADA on `preprod`, you can use the [Testnet Faucet](https://docs.cardano.org/cardano-testnet/tools/faucet/) to fund your wallet or the addresses above. Note that due to rate limiting, it's better to request large sums for efficiency and distribute as needed.

Reviewer: ```suggestion
In case you don't have test ada on `pre-production`, you can use the [testnet faucet](https://docs.cardano.org/cardano-testnets/tools/faucet/) to fund your wallet or the addresses above. Note that due to rate limiting, it's better to request large sums for efficiency and distribute as needed.
```

```

</TabItem>

> </Tabs>

Reviewer 2: [CI](https://github.com/input-output-hk/hydra/actions/runs/9685477231/job/26725955291?pr=1469) fails because of this deletion.
Should revert this change to make it pass.

Reviewer 2: There were a bunch of errors happening, and this was the only one getting reported.
The main issue was that even after fixing this error, the same report continued to appear, obscuring the actual underlying causes.

Shame on Docusaurus!

</TabItem>
</Tabs>

> Next, generate Hydra key pairs for use on layer two. Use the

Reviewer: ```suggestion
Next, generate Hydra key pairs for use on layer 2. Use the
```

Besides keys, we also want to communicate each other's connectivity information.
That is, an IP address / hostname + port where we will be reachable for our
layer two network using `hydra-node`. For the purpose of this tutorial we are
assuming an IP address and port for `alice` and `bob` which works on a single
machine, but please replace usages below with your respective addresses:
Before launching the `hydra-node`, it's crucial to establish and communicate each participant's network connectivity details. This includes the IP addresses and ports where `Alice` and `Bob's` nodes will be reachable for the layer two network interactions. For this tutorial, we're using placeholder IP addresses and ports which should be replaced with your actual network details:

<!-- TODO: can we make peers configurable via some text input? -->

Alice: <code>127.0.0.1:5001</code>

> Bob: <code>127.0.0.1:5001</code>

Reviewer: Should this line be deleted, @oduameh?

Author: Thanks for flagging this @olgahryniuk I accidentally deleted this line

</TabItem>
</Tabs>

> If you are collaborating with a partner, exchange the verification (public) keys: `{alice,bob}-node.vk` and `{alice,bob}-hydra.vk` to ensure secure communication

Reviewer: ```suggestion
If you are collaborating with another individual, exchange the verification (public) keys: `{alice,bob}-node.vk` and `{alice,bob}-hydra.vk` to ensure secure communication.
```

> Before launching the `hydra-node`, it's crucial to establish and communicate each participant's network connectivity details. This includes the IP addresses and ports where `Alice` and `Bob's` nodes will be reachable for the layer two network interactions. For this tutorial, we're using placeholder IP addresses and ports which should be replaced with your actual network details:

Reviewer: ```suggestion
Before launching the `hydra-node`, it's crucial to establish and communicate each participant's network connectivity details. This includes the IP addresses and ports where `Alice` and `Bob's` nodes will be reachable for layer 2 network interactions. For this tutorial, we're using placeholder IP addresses and ports which should be replaced with your actual network details:
```

<!-- TODO: can we make peers configurable via some text input? -->

Alice: <code>127.0.0.1:5001</code>

Bob: <code>127.0.0.1:5001</code>

> The next step involves configuring the protocol parameters for the ledger within our Hydra head. For the purposes of this tutorial, we'll modify the default Cardano layer one parameters to eliminate transaction fees, simplifying our test interactions:

Reviewer: ```suggestion
The next step involves configuring the protocol parameters for the ledger within our Hydra head. For the purposes of this tutorial, we'll modify the default Cardano layer 1 parameters to eliminate transaction fees, simplifying test interactions:
```

```
cardano-cli query protocol-parameters \
  | jq '.txFeeFixed = 0 |.txFeePerByte = 0 | .executionUnitPrices.priceMemory = 0 | .executionUnitPrices.priceSteps = 0' \
  > protocol-parameters.json
```

> This command adjusts the fees and pricing mechanisms to zero, ensuring that transactions within our Hydra head incur no costs.

Reviewer: ```suggestion
This command adjusts the fees and pricing mechanisms to zero, ensuring that transactions within the Hydra head incur no costs.
```

In summary, the Hydra head participants exchanged and agreed on:

> - IP addresses + port on which their `hydra-node` will run.

Reviewer: Suggest updating the list thus:

- IP addresses and the port on which their `hydra-node` will run
- A Hydra verification key to identify them in the head
- A Cardano verification key to identify them on the blockchain
- Protocol parameters to use in the Hydra head
- A contestation period for the head closing (we will use the default here).

Author: Updated

> ## Step 3: Start the Hydra node

Reviewer: ```suggestion
## Step 3. Start the Hydra node
```

With the necessary parameters defined, we are now ready to initiate the Hydra nodes. First, select the version of the Hydra Head protocol you wish to use. This version corresponds to the `hydra-node --HYDRA_VERSION` and the `--hydra-scripts-tx-id` which refers to scripts published on-chain.

> For all [released](https://github.com/input-output-hk/hydra/releases) HYDRA_VERSIONs of the `hydra-node` and common Cardano networks,  scripts are pre-published. Consult the [user manual](../configuration#reference-scripts) for guidance on publishing your own scripts.

Reviewer: ```suggestion
Scripts are pre-published for all [released](https://github.com/input-output-hk/hydra/releases) HYDRA_VERSIONs of the `hydra-node` and common Cardano networks. Consult the [user manual](../configuration#reference-scripts) for guidance on publishing your own scripts.
```

}
```

Ensure that both nodes are connected before proceeding. If the `PeerConnected` message does not appear, review your network settings and ensure connectivity is established.

> ## Step 4: Open a Hydra head

Reviewer: ```suggestion
## Step 4. Open a Hydra head
```

}
```

> ## Step 4. Open a Hydra head

Reviewer 3: This is a duplicate headline to the one below!

Also: do we want to really change the headline from `:` to `.`? If yes, all of them should change consistently.

Reviewer: Our style guidelines typically follow the format 'Step x. Start with a capital' or 'Step x: use lowercase', with a preference for the first option. I believe this line is the only inconsistency. @oduameh, please address this by deleting one of the duplicated headlines (the one with `:`).

Author: Updated globally

{ "tag": "Init" }
```

The initiation process might take some time as it includes submitting a transaction on-chain. Upon successful initiation, both Hydra nodes and their clients will display a `HeadIsInitializing` message, listing the parties required to commit.

<!-- TODO: Add troubleshooting here + abort and checking keys -->

> To commit funds to the head, meaning to choose which UTXOs to make available on layer two. Use the HTTP API of `hydra-node` to commit all

Reviewer: ```suggestion
To commit funds to the head, you need to choose which UTXOs you would like to make available on layer 2. Use the HTTP API of `hydra-node` to commit all
```

</details>

> After you've prepared your transactions, the `hydra-node` will find all UTxOs associated with the funds key and create a draft of the commit transaction. You'll then sign this transaction using the funds key and submit it to the Cardano layer one network. 

Reviewer: ```suggestion
After you've prepared your transactions, the `hydra-node` will find all UTXOs associated with the funds key and create a draft of the commit transaction. You'll then sign this transaction using the funds key and submit it to the Cardano layer 1 network. 
```

> Once this transaction is seen by the `hydra-node`, you should see a `Committed` status will be displayed on your WebSocket connection.
> 
> When both parties, `alice` and `bob`, have committed,  the Hydra head will open automatically. You'll see a `HeadIsOpen` message appear in the WebSocket session, confirming the activation of the head. This message will include details such as the starting balance and UTXO entries. Notably, these entries will match exactly those committed to the head, including transaction hashes and indices, ensuring transparency and consistency.
> 
> The head is now operational and ready for further activities.

Reviewer: ```suggestion
Once the `hydra-node` sees this transaction, you should see a `Committed` status displayed on your WebSocket connection.

When both parties, `alice` and `bob`, have committed,  the Hydra head will open automatically. You'll see a `HeadIsOpen` message appear in the WebSocket session, confirming the activation of the head. This message will include details such as the starting balance and UTXO entries. Notably, these entries will match exactly those committed to the head, including transaction hashes and indices, ensuring transparency and consistency.

The head is now operational and ready for further activities.
```

> ## Step 5: Using the Hydra head

Reviewer: ```suggestion
## Step 5. Use the Hydra head
```

> In this step, we'll demonstrate a basic transaction between `alice` and `bob` using the Hydra head. Hydra Head operates as an isomorphic protocol, meaning that functionalities available on the Cardano layer one network are also available on the layer two network. This compatibility allows us to use familiar tools like `cardano-cli` for transaction creation within the head.

Reviewer: ```suggestion
In this step, we'll demonstrate a basic transaction between `alice` and `bob` using the Hydra head. Hydra Head operates as an isomorphic protocol, meaning that functionalities available on the Cardano layer 1 network are also available on the layer 2 network. This compatibility allows us to use familiar tools like `cardano-cli` for transaction creation within the head.
```

> In this example, we will transfer `10₳` from Alice to Bob. Adjust the transaction amount based on the balances previously committed to the head.

Reviewer: ```suggestion
In this example, we will transfer 10 ada from Alice to Bob. Adjust the transaction amount based on the balances previously committed to the head.
```

  --out-file tx-signed.json
```

Submit the transaction through the already open WebSocket connection. Generate the `NewTx` command for WebSocket submission:

```shell
cat tx-signed.json | jq -c '{tag: "NewTx", transaction: .}'
```

The transation will be validated by both `hydra-node`s and either result in a
`TxInvalid` message with a reason, or a `TxValid` message and a
`SnapshotConfirmed` with the new UTXO available in the head shortly after.

🎉 Congratulations, you just processed your first Cardano transaction off-chain
in a Hydra head!

Feel free to continue experimenting with additional transactions. Move on to the next steps when ready to finalize your activities within the head.

> ## Step 6: Closing the Hydra head

Reviewer: ```suggestion
## Step 6. Close the Hydra head
```

Any participant can initiate closing the Hydra head. Use the WebSocket API to submit the closing command:

```json title="Websocket API"

> { "tag": "Close" }

Reviewer: Please update the copy below thus:

The `hydra-node` will then submit a protocol transaction with the last known snapshot to the Cardano network. A smart contract on layer 1 will check the snapshot signatures and confirm the head is closed. The WebSocket API sends a `HeadIsClosed` message when this' Close' transaction is observed. Note that this can also happen if any other `hydra-node` closes the head.

:::caution Known bug

You might need to submit the `Close` command multiple times if the head is not getting closed within approximately 30 seconds.

See [#1039](https://github.com/input-output-hk/hydra/issues/1039) for details.

:::

The message will include a `contestationDeadline`, set using the configurable `--contestation-period`. Until this deadline, the closing snapshot can be contested with a more recent, multi-signed snapshot. Your `hydra-node` would contest automatically for you if the closed snapshot is not the last known one.'

Author: Updated
